exercises/joseph_torsney/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,161 @@
+#include <iostream> 
+#include <fstream>
+#include <vector>
+#include <map>
+#include <algorithm>
+
+/** Checks whether a character is punctuation
+ * 
+ * @param c 
+ * @return true if c is punctuation
+ * @return false otherwise
+ */
+bool myIsPunct(char c) {
+  return(
+    c == '.' ||
+    c == ',' ||
+    c == '?' ||
+    c == '\'' ||
+    c == '\"' ||
+    c == '!' ||
+    c == ')' ||
+    c == '('
+  );
+}
+
+/** Reads the contents of a .txt file at filepath
+ * 
+ * @param filepath the path to the file
+ * @return std::string of the file contents
+ */
+std::string readFile(const std::string filepath)
+{
+  std::ifstream ifs(filepath);
+  std::string text(
+    (std::istreambuf_iterator<char>(ifs)),
+    (std::istreambuf_iterator<char>())
+  );
+  
+  return(text);
+}
+
+/** Removes punctuation from a string
+ * 
+ * @param str the string to process
+ * @return std::string with no punctuation
+ */
+std::string removePunct(std::string str)
+{
+  for (int i = 0, len = (int) str.size(); i < len; i++) {
+    if (myIsPunct(str[i])) {
+      // if this char is punctuation, reduce i by 1, remove it, and
+      // reevaluate the string size.
+      str.erase(i--, 1);
+      len = (int) str.size();
+    }
+  }
+  return(str);
+}
+
+/** Splits a string by spaces and hyphens into a string vector.
+ * 
+ * @param s the string to split
+ * @return std::vector<std::string> of 'words' contained in s.
+ */
+std::vector<std::string> split(const std::string& s)
+{
+  std::vector<std::string> ret;
+  typedef std::string::size_type string_size;
+  string_size i = 0;
+
+  // process characters of s
+  while (i != s.size()) {
+
+    // ignore whitespace
+    while(i != s.size() && (isspace(s[i]) || s[i] == '-')) {
+      i++;
+    }
+
+    // otherwise, we are at the start of a word.
+    string_size j = i;
+
+    // find the position of the next space.
+    while (j != s.size() && !(isspace(s[j]) || s[j] == '-')) {
+      j++;
+    }
+
+    // If we have found a word
+    if (i != j) {
+      // copy substring from i, taking j - i chars.
+      ret.push_back(s.substr(i, j - i));
+      i = j;
+    }
+  }
+
+  return(ret);
+}
+
+/** Counts the number of times a word appears (usage) in a given string vector.
+ * 
+ * @param words the string vector to count
+ * @return std::map<std::string, int> maps each word to the number of usages.
+ */
+std::map<std::string, int> countWords(std::vector<std::string> words) {
+  std::map<std::string, int> usage;
+  std::map<std::string, int>::iterator it;
+
+  for (int i = 0; i < words.size(); i++) {
+    it = usage.find(words[i]);
```

Comment:
I think you can replace the contents of this loop by simply:
usage[words[i]]++
I think the map manages whether the entry exists or not for you

Reply:
Don't think you've made this change yet? Just in case I wasn't clear I think you can just do:
```
  for (int i = 0; i < words.size(); i++) {
    usage[words[i]]++;
  }
```

Reply:
Thought I had but I was messing around with git so that's probably why!